Elastic Bulk Document Creation

elipe17 opened the pull request:
## Summary of Changes
- Fixed issue where Elastic documents were not being created due to Django bulk model creation not sending appropriate signals
- Bypassed Django signals and manually bulk create Elastic documents for minimized I/O
- This also includes the addition of the Kibana container

## How to Test

```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up --build
```

- Open http://localhost:3000/ and sign in.
- Submit any number of data files
- Verify documents exist in Elastic by running the following command in your terminal: `curl -H "Content-Type: application/json" -X GET "localhost:9200/<INDEX_NAME>/_count"` where <INDEX_NAME> can be `tanf_t1_submissions`.
- Another good verification would be to submit `ADS.E2J.NDM1.TS53_fake.txt` and verify that it takes minutes and not hours to parse/validate. Should be around 3 - 5 minutes.


## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Parsing is still as fast as possible
+ [x] Document creation actually happens now
+ [ ] `adpennington` approved

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

raftmsohani committed:
Added formating for header and autofit columns

raftmsohani committed:
Formatted the headers

raftmsohani committed:
added year/month to the columns

raftmsohani committed:
Added contants - translation column

raftmsohani committed:
added friendly names to T1 and T2

raftmsohani committed:
added friendly name to m1 and m2

raftmsohani committed:
added friendly name to m3

raftmsohani committed:
added friendly_name to t3

raftmsohani committed:
added friendly_name to t4 and t5

raftmsohani committed:
added friendly_name to t7

raftmsohani committed:
correct missing friendly_name

raftmsohani committed:
correction on failing tests

raftmsohani committed:
addedfriendly name to excel report

raftmsohani committed:
linting

raftmsohani committed:
linting

raftmsohani committed:
linting

raftmsohani committed:
delete contants.py

raftmsohani committed:
added test for json field in error model

raftmsohani committed:
linting

raftmsohani committed:
linting

raftmsohani committed:
linting

raftmsohani committed:
2599-added friendly name to postparsing validators

raftmsohani committed:
refining the validator tests

raftmsohani committed:
added returning fields names to validators

raftmsohani committed:
added friendly_name to error field

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
linting

raftmsohani committed:
corrections on views/tests

raftmsohani committed:
corrections for fields

raftmsohani committed:
failing test corrected

raftmsohani committed:
failing test corrected

raftmsohani committed:
correcting test failures

raftmsohani committed:
linting

raftmsohani committed:
corrected the excel fiel generator

raftmsohani committed:
removed excessive space in validator

raftmsohani committed:
linting

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
listing

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
added m6

raftmsohani committed:
lint

raftmsohani committed:
corrected new line break

raftmsohani committed:
refactored validator logic

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
linting and correction on t1

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
friendly_name correction from comments

raftmsohani committed:
Merge branch '2599-error-report-corrections-impact-wpr-metrics' of https://github.com/raft-tech/TANF-app into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
friendly_name correction

raftmsohani committed:
corrected failing test for m5

raftmsohani committed:
refactor the field_json creation DRY

Commit:
Changes for fully local development
 - Enables direct frontend/backend communication sans Login.gov/Cloud.gov
 - Drives off new DEVELOPMENT env var
 - Pre-configures and disables frontend auth functionality
 - Testing based on new dev user
   - Install via web: ./manage.py generate_dev_user

elipe17 committed:
- Added Kibana config

raftmsohani committed:
friendly_name corrections

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

elipe17 committed:
Merge branch '2599-error-report-corrections-impact-wpr-metrics' of https://github.com/raft-tech/TANF-app into kibana-local

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into kibana-local

raftmsohani committed:
linting and cleaning errors

raftmsohani committed:
linting

raftmsohani committed:
correction on friendly_names

raftmsohani committed:
corrected friendly_name for test_util

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
correction child care - number of months

raftmsohani committed:
Merge branch '2599-error-report-corrections-impact-wpr-metrics' of https://github.com/raft-tech/TANF-app into 2599-error-report-corrections-impact-wpr-metrics

Commit:
Reorganized front end logic on REACT_APP_DEVAUTH env var

Commit:
Reorganized backend logic on REACT_APP_DEVAUTH env var

George-Hudson committed:
fixed a few more typos and some spacing. (#2767)

* fixed a few more typos and some spacing.

* fixed linting issues

* missed a spot.

---------

Co-authored-by: George Hudson <ghudson@teamraft.com>

elipe17 committed:
- Added basic security to Kibana/Elastic
- Added setup container to init elastic users, roles, and passwords

elipe17 committed:
- Remove debug code

elipe17 committed:
- change provider name

elipe17 committed:
- Updating settings to reference environment variables

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1858-kibana-ofa-access

elipe17 committed:
- Add elastic dependency

elipe17 committed:
- Fix network issue

elipe17 committed:
- Added bulk creation of elastic indices

elipe17 committed:
- Updated schemas to reference model based off of elastic document

elipe17 committed:
- Remove password auth from elastic/kibana

elipe17 committed:
- Fix tests

elipe17 committed:
- Fix lint

andrew-jameson commented:
Will postpone reviewing this until the base branch issue is resolved given the extra chaff in this PR.

elipe17 committed:
- remove debug print

elipe17 committed:
Merge branch '2599-error-report-corrections-impact-wpr-metrics' of https://github.com/raft-tech/TANF-app into elastic-bulk-doc-creation

Commit:
added is_superuser and is_staff attrs to dev user

elipe17 committed:
- Fix lint errors

tptignor commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -55,22 +55,27 @@ def parse_datafile(datafile):
-                num_created += len(model.objects.bulk_create(records))
-            if num_created != num_expected:
-                logger.error(f"Bulk create only created {num_created}/{num_expected}!")
+            num_records_created = 0
```
Hi Eric. I've been rereading this over and over... The code looks right to me. All my comments are on clarity. First, as you may know, an ES "document" is synonymous with a record in this context. ("Documents are JSON objects that are stored within an Elasticsearch index and are considered the base unit of storage. In the world of relational databases, documents can be compared to a row in table.") So you could name all the variables as "records" (or "recs"). Also, it seems we're tracking creates in three resources. I think it would help to name the resources, so maybe something like: 

- num_expected_records --> postgres_records
- num_records_created --> django_records
- num_documents_created --> elastic_records

tptignor commented on tdrs-frontend/docker-compose.yml:
```diff
@@ -7,7 +7,7 @@ services:
       - 8090:8090
     networks:
-      - local
+      - default
```
I'd like to invite others to comment on this change. I removed the local network because I found it obstructing container comm in my local environment. It must have been put here for a reason, though. Can someone provide insight on the local and external-net networks and how they were meant to be used by frontend and backend containers?

tptignor commented on tdrs-frontend/src/configureStore.js:
```diff
@@ -13,9 +14,29 @@ export const history = createBrowserHistory()
   const middlewares = [thunkMiddleware, loggerMiddleware]
   const composedEnhancers = composeWithDevTools(applyMiddleware(...middlewares))
+  const devState = {
+    router: { location: { pathname: '/profile' } },
```
I believe a few additional changes will be needed for the DEVAUTH feature. Hoping to get some time with @jtimpe today to figure this out.

elipe17 replied:
@tptignor does it make sense to move the devauth stuff to a separate PR if there is still going to be active work on it? That way this PR isn't blocked and the elastic indexes/documents can be created in the deployed envs?

tptignor replied:
Hi guys. I actually put a ton of work in the last 12 hours into rewriting and verifying devauth. What's here now should be good to go, and it passes the unit tests that were passing before. However, if someone wants to recommend a different base branch to port it to, I completely understand and will get to that as soon as I can. (Need to get back to DIDDMS work now...)

jtimpe replied:
@tptignor @elipe17 i'd like to see these PRs split up just on the basis of having clean diffs and commit histories when we go to merge them. Both sets of changes currently look good to me though

elipe17 replied:
Pulled Dev Auth out of this PR.

elipe17 committed:
- Renaming variables to clarify things

elipe17 committed:
- fix lint

Commit:
DevAuth feature redesign inspired by Cypress
 - Initializing frontend w/POST /login/cypress: {devEmail, local-cypress-token}
 - Changed REACT_APP_DEVAUTH to provide the email of the desired dev user
 - Modified CustomAuthentication.authenticate to handle both known use cases
 - Added stt_id=31 to the initial dev user
 - Disabled ES disk threshold checking for local dev which blocked ES startup
 - Removed DevAuthentication and other now unnecessary code

Commit:
Merge branch 'elastic-bulk-doc-creation' of https://github.com/raft-tech/TANF-app into elastic-bulk-doc-creation
Merging latest changes from server

Commit:
Fixed CustomAuthentication.authenticate return val for login.py use case

Commit:
Fixed CustomAuthentication.authenticate logging for login.py use case

Commit:
Removed unneeded permissions import

Commit:
Updates to REACT_APP_DEVAUTH env var settings
 - Enabled with an email address value
 - Disabled by default

tptignor commented:
@elipe17 : Your changes yesterday look great. Thanks!

Commit:
Restored support for CustomAuthentication.authenticate username keyword

Commit:
Modified CustomAuthentication.authenticate comment to satisfy flake8

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -55,22 +55,29 @@ def parse_datafile(datafile):
+                num_elastic_records_created += document.update(created_objs)[0]
+                num_db_records_created += len(created_objs)
+            if num_db_records_created != num_expected_db_records:
+                logger.error(f"Bulk Django record creation only created {num_db_records_created}/" +
```
Suggestion is to add line_number/header count (if it is meaningful) for understanding the error and being able to debug from the log

jtimpe commented on tdrs-frontend/src/index.js:
```diff
@@ -24,8 +24,28 @@ axios.defaults.xsrfCookieName = 'csrftoken'
+    .catch(function (error) {
+      console.log(error)
+    })
+  store.dispatch({ type: 'SET_AUTH', payload: { devEmail } })
```
i think the `payload` passed here needs to be changed to `payload: { email: devEmail }`

what you currently have is javasscript shorthand for `{ devEmail: devEmail }`

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into elastic-bulk-doc-creation

elipe17 committed:
commit

elipe17 committed:
Revert "Modified CustomAuthentication.authenticate comment to satisfy flake8"

This reverts commit 761e4eb253d366ef742dd8caf94b6220ed9e81a1.

elipe17 committed:
Revert "Restored support for CustomAuthentication.authenticate username keyword"

This reverts commit 4bf895722e356e79b8bbe3674361b90888b91752.

elipe17 committed:
Revert "Updates to REACT_APP_DEVAUTH env var settings"

This reverts commit 7fc2a09353804fb728852e9accc041dff08e44e3.

elipe17 committed:
Revert "Removed unneeded permissions import"

This reverts commit c18383fe2bf8352c50dd84d2a84408fef2b71367.

elipe17 committed:
Revert "Fixed CustomAuthentication.authenticate logging for login.py use case"

This reverts commit 2b9b46f0e719638b320d0a7bbb2bd87eda97eeef.

elipe17 committed:
Revert "Fixed CustomAuthentication.authenticate return val for login.py use case"

This reverts commit 97a0cf6995dc17937f083d5efde1d74866c01ff4.

elipe17 committed:
Revert "DevAuth feature redesign inspired by Cypress"

This reverts commit 1497d4ab7549bf674e1f71d8f8f039ec7de363bf.

elipe17 committed:
Revert "commit"

This reverts commit a284856c66cb2caaca4b471c2f81fc39421e8f00.

elipe17 committed:
Revert "added is_superuser and is_staff attrs to dev user"

This reverts commit 6ffbee8f588f12b7595abd7adcfe15ad7e70d11b.

elipe17 committed:
Revert "Reorganized backend logic on REACT_APP_DEVAUTH env var"

This reverts commit 7fd7b4d48cd30a7e2c142ea7f1d85f8df95b80d8.

elipe17 committed:
Revert "Reorganized front end logic on REACT_APP_DEVAUTH env var"

This reverts commit 32a46713ae102fa15fdab0fcb66fca99b42eb7e2.

elipe17 committed:
Revert "Changes for fully local development"

This reverts commit 556221b310b73bee5e9af32eb4cd603b61a25d02.

elipe17 committed:
asdf

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -107,7 +114,8 @@ def evaluate_trailer(datafile, trailer_count, multiple_trailer_errors, is_last_l
-    for model in unsaved_records:
+    for document in unsaved_records:
+        model = document.Django.model
         num_deleted, models = model.objects.filter(datafile=datafile).delete()
```
out of curiosity i ran the big `fake_rollback` file and noticed the delete requests going crazy

https://github.com/raft-tech/TANF-app/assets/111305129/302b5565-3979-44c7-b776-0fb0505feaf4

but the db and elastic eventually both end up in a consistent state.

this may be a topic for another ticket, but implementing some sort of `bulk_delete` may be helpful. existing solutions might not honor the signals needed for elastic, though. https://stackoverflow.com/a/36935536

elipe17 replied:
Agreed, our transaction-less rollback is a pain point for sure. It would be interesting to test out the `raw_delete`.

raftmsohani replied:
have to be careful with ```raw_delete``` though!

jtimpe commented on tdrs-backend/tdpservice/parsers/row_schema.py:
```diff
@@ -12,13 +12,13 @@ class RowSchema:
+            quiet_preparser_errors=False,
             ):
-        self.model = model
+        self.document = document
```
this change makes me nervous because of the potential side-effects, but you've covered all the cases i've thought of so far

jtimpe reviewed (approved):
Good catch and nice fix!

George-Hudson reviewed (approved):
LGTM

elipe17 committed:
- Adding integration tests for elastic bulk doc creation

elipe17 committed:
Revert "asdf"

This reverts commit 26455b48582ca9c6d986377e56475644525f7665.

elipe17 committed:
- fix lint

elipe17 committed:
fasdf

jtimpe commented on tdrs-backend/tdpservice/search_indexes/documents/document_base.py:
```diff
@@ -1,16 +1,18 @@
 
     datafile = fields.ObjectField(properties={
-                      'pk': fields.IntegerField(),
+                      'id': fields.IntegerField(),
```
only thing i would say here is that `id` is not always the `pk` and we've run into cases in the past where this assumption has broken down. everything works in this case, though

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into elastic-bulk-doc-creation

elipe17 committed:
- Added usage of document to tribal

ADPennington reviewed (approved):
@elipe17 test results below. im not seeing the large file fully parse locally or in dev environment, so im unable to check those bulk create results. i assume thats beyond the scope of this ticket, so no reason to hold this up!


```shell
apennington@HHSLBDSWL73 MINGW64 ~/GitHub/RAFT/TANF-app (elastic-bulk-doc-creation)
$ curl -H "Content-Type: application/json" -X GET "localhost:9200/tanf_t1_submissions/_count"

{"count":1,"_shards":{"total":1,"successful":1,"skipped":0,"failed":0}}

# large file parsing timed out locally and apps crashed, potentially due to memory issues
```

dev env results for large file submission

```shell

2024-01-11T16:17:00.42-0500 [APP/PROC/WEB/0] ERR [2024-01-11 21:17:00,425: ERROR/MainProcess] Process 'ForkPoolWorker-2' pid:173 exited with 'signal 9 (SIGKILL)'
   2024-01-11T16:17:00.43-0500 [APP/PROC/WEB/0] ERR [2024-01-11 21:17:00,437: ERROR/MainProcess] Task handler raised error: WorkerLostError('Worker exited prematurely: signal 9 (SIGKILL) Job: 2.')
   2024-01-11T16:17:00.43-0500 [APP/PROC/WEB/0] ERR Traceback (most recent call last):
   2024-01-11T16:17:00.43-0500 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/billiard/pool.py", line 1265, in mark_as_worker_lost
   2024-01-11T16:17:00.43-0500 [APP/PROC/WEB/0] ERR raise WorkerLostError(
   2024-01-11T16:17:00.43-0500 [APP/PROC/WEB/0] ERR billiard.exceptions.WorkerLostError: Worker exited prematurely: signal 9 (SIGKILL) Job: 2.

```

andrew-jameson committed:
Merge branch 'develop' into elastic-bulk-doc-creation

elipe17 committed:
- Fixing error

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into elastic-bulk-doc-creation
